src/eredis_sentinel.erl

```diff
@@ -0,0 +1,251 @@
+    {stop, Reason :: term()} | ignore).
+init(Options) ->
+    process_flag(trap_exit, true),
+    MasterGroup         = proplists:get_value(master_group, Options, mymaster),
+    SentinelAddresses   = proplists:get_value(addresses, Options, [{"127.0.0.1", 26379}]),
+    Username            = proplists:get_value(username, Options, undefined),
+    Password            = proplists:get_value(password, Options, ""),
+    ConnectTimeout      = proplists:get_value(connect_timeout, Options, ?CONNECT_TIMEOUT),
+    SocketOptions       = proplists:get_value(socket_options, Options, []),
+    TlsOptions          = proplists:get_value(tls, Options, []),
+    {ok, #eredis_sentinel_state{master_group = MasterGroup,
+        addresses = SentinelAddresses,
+        username = Username,
+        password = Password,
+        connect_timeout = ConnectTimeout,
+        socket_options = SocketOptions,
+        tls_options = TlsOptions,
+        conn_pid = undefined,
+        errors    = #errors{}}}.
+
+%% @private
+%% @doc Handling call messages
+-spec(handle_call(Request :: term(), From :: {pid(), Tag :: term()},
+    State :: #eredis_sentinel_state{}) ->
+    {reply, Reply :: term(), NewState :: #eredis_sentinel_state{}} |
+    {reply, Reply :: term(), NewState :: #eredis_sentinel_state{}, timeout() | hibernate} |
+    {noreply, NewState :: #eredis_sentinel_state{}} |
+    {noreply, NewState :: #eredis_sentinel_state{}, timeout() | hibernate} |
+    {stop, Reason :: term(), Reply :: term(), NewState :: #eredis_sentinel_state{}} |
+    {stop, Reason :: term(), NewState :: #eredis_sentinel_state{}}).
+handle_call(get_master, _From, State) ->
+    case query_master(State#eredis_sentinel_state{errors = #errors{}}) of
+        {ok, {Host,Port}, S1} ->
+            {reply, {ok, Host,Port}, S1};
+        {error, Error, S1} ->
+            {reply, {error, Error}, S1}
+    end;
+
+handle_call(stop, _From, State) ->
+    {stop, normal, ok, State}.
+
+%% @private
+%% @doc Handling cast messages
+-spec(handle_cast(Request :: term(), State :: #eredis_sentinel_state{}) ->
+    {noreply, NewState :: #eredis_sentinel_state{}} |
+    {noreply, NewState :: #eredis_sentinel_state{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #eredis_sentinel_state{}}).
+handle_cast(_Request, State = #eredis_sentinel_state{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc Handling all non call/cast messages
+-spec(handle_info(Info :: timeout() | term(), State :: #eredis_sentinel_state{}) ->
+    {noreply, NewState :: #eredis_sentinel_state{}} |
+    {noreply, NewState :: #eredis_sentinel_state{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #eredis_sentinel_state{}}).
+%% Current sentinel connection broken
+handle_info({'EXIT', Pid, _Reason}, #eredis_sentinel_state{conn_pid = Pid} = S) ->
+    {noreply, S#eredis_sentinel_state{conn_pid = undefined}};
+
+%% Ignore late exit messages
+handle_info({'EXIT', _Pid, _Reason}, S) ->
+    {noreply, S};
```

Comment:
If the `eredis_sentinel` process is linked to an `eredis_client` process which started the sentinel process (start_link) and the `eredis_client` process crashes, the `eredis_sentinel` process survives and is leaked.

It is risky to use `trap_exit` without a supervisor. I don't think it should be done. A supervisor can handle a process which ignores the 'EXIT' message by using brutal kill after a timeout, so it is the safe way to do it, or don't use trap exit if it's not needed. Can monitor be used instead?

Reply:
So there are two ways. 
Since monitor is a unidirectional link, we'll need to pass eredis_sentinel process Pid to sentinel connection process so that it can setup a monitor link to avoid leak when sentinel crashes.
Or maybe we can keep trap_exit and stop the process on receiving exit except when it is from sentinel connection process by simply changing line 127 to `{stop, normal, S}`.

Reply:
You're right about monitor. I think we can't use it.

Trapping exit and handling all exit signals with `{stop, Reason, S}` would probably work.

However, handling processes like this is a bit like reinventing a supervisor. During the Code BEAM conference I talked to some people in the OTP team. They say that you should always use supervisors when you use multiple processes and links. The supervisor module is more robust and more tested. Reinventing the supervisor should not be done, because there is a risk of bugs, now or in code added later.

But how can we design a supervision tree for this? (I don't know.... Maybe we would need to use an `application` behaviour and add all eredis_sentinel processes to the eredis application supervisor? But starting the application is only useful when sentinel is used. Normally, eredis is not started as a separate application, but as part of the user's supervision trees. Maybe we can let `eredis_sentinel` to be an application, so it is only started when sentinel is used?....)

I think we can use trap exit it for now. We can say we have *experimental support* for sentinel, and maybe later we can change to use supervisors. What do you think?

Reply:
As a first step to add sentinel support, it seems fine to me as well.